File: exercises/floor01/project_character.py
"""
FLOOR 1 · PROJECT — THE ROLL (character creator)
This is the first piece of your final roguelike. Everything you build in
later floors will act on the character dict you create here.

Functions get their full lesson on Floor 4 — for now, follow the recipe:
the skeletons are written, you fill in the bodies. A function takes inputs
(parameters), does work, and hands back a result with `return`.

Run:  python3 dojo.py check 1-P
Save: python3 dojo.py done 1-P
"""

STARTING_HP = 30
STARTING_LEVEL = 1
STARTING_GOLD = 10


def make_character(name):
    """Build and return the character dict.

    Return a dict with exactly these keys:
        "name"  -> the name that was passed in
        "hp"    -> STARTING_HP
        "max_hp"-> STARTING_HP
        "level" -> STARTING_LEVEL
        "gold"  -> STARTING_GOLD

    Use the CONSTANTS above, not raw numbers — when you rebalance the game
    later, you change one line instead of ten.
    """
    # TODO: build the dict and return it
    ...


def status_line(character):
    """Return the HUD string for a character dict, EXACTLY in this shape:

        Crumb | Lv 1 | HP 30/30 | 10g

    (using the character's actual values, via an f-string)
    Dict values are read with square brackets: character["name"]
    """
    # TODO
    ...


def level_up(character):
    """Level the character up, then return the (same) dict.

    Leveling up means:
        level  +1
        max_hp +5
        hp     restored to the new max_hp   (ding heals — classic RPG rule)
    """
    # TODO
    ...


if __name__ == "__main__":
    # Play with your creation! `python3 exercises/floor01/project_character.py`
    hero = make_character("Crumb")
    print(status_line(hero))
    level_up(hero)
    print(status_line(hero))

File: exercises/cpp/floor01/project_character.cpp
// FLOOR C1 · PROJECT — THE ROLL (character creator, C++ edition)
// The same first organ you built in Python — now with a struct instead of a
// dict, and a preview of references (&). Follow the recipe; structs get
// their full lesson later.
//
// Run:  python3 dojo.py check c1-P
// Save: python3 dojo.py done c1-P

#include <string>

const int STARTING_HP = 30;
const int STARTING_LEVEL = 1;
const int STARTING_GOLD = 10;

// A struct is a bundle of named fields — the C++ answer to your Python dict,
// except the fields and their types are fixed at compile time.
struct Character {
    std::string name;
    int hp;
    int max_hp;
    int level;
    int gold;
};

// Build and return a Character using the constants above.
// Struct fields are set with dot syntax:  c.hp = STARTING_HP;
// (or brace-init in one line: return {name, STARTING_HP, ...} — field order!)
Character make_character(std::string name) {
    // TODO
    return {};
}

// Return EXACTLY:  Crumb | Lv 1 | HP 30/30 | 10g
// (std::to_string for every number)
std::string status_line(Character c) {
    // TODO
    return "";
}

// Level up the character: level +1, max_hp +5, hp restored to new max.
// NOTE the &: Character& means "the actual character, not a copy."
// Without it you'd level up a copy, the original stays weak, and you'd
// hunt that bug for an hour. (Python's dicts behaved this way for free.)
void level_up(Character& c) {
    // TODO
}
